### PR TITLE
Update AIControlConfig to use dependency injection

### LINKED ISSUE #1137: Dependency Inject AI Control parameters
### Description of the task

See #960 for more details

The `AI Control` parameters should be changed to use dependency injection, by passing in the `AI Control` objects to places that need it. (tip: `grep` or search for usages of `AIControl` to find where the data config needs to go). Replace and remove the singleton usages of the parameters (they look like `Util::DynamicParameters::etc`)

### Acceptance criteria

- [ ] Inject the AI Control parameters into all places needed

### PR DESCRIPTION
**Pull request by LiCody**
### Description

AIControlConfig now uses dependency injection. AIControlConfig is now passed through the constructor of several objects.

### Testing Done
Builds only tests affected were in stp_test and grsim_output_test

### Resolved Issues
resolves #1137 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/visualizer/widgets/main_widget.cpp
Comment (HaomiaoZ) [inline src/software/visualizer/widgets/main_widget.cpp:5]: Just a question here, why are we passing the config here if it is not assigned to a member variable?

**Review by HaomiaoZ — CHANGES_REQUESTED** [review]
Overall LGTM. I do have a question commented separately. I am not too sure why build fails based on the changes.

**Comment by garethellis0** [comment]
Just a note, when you're indicating issues associated with this PR, if you write them as `resolves #1234` instead of just `#1234` it will close the issue automatically when the PR gets merged.

File: src/software/ai/ai.h
Comment (garethellis0) [inline src/software/ai/ai.h:25]: We should update the javadoc comment here as well

File: src/software/ai/hl/stp/stp.cpp
Comment (garethellis0) [inline src/software/ai/hl/stp/stp.cpp]: Not sure why we added this include? If we need to include `shared_ptr`, we should use `#include <memory>` as `utility` contains more things that we probably don't need here.

File: src/software/backend/grsim_backend.cpp
Comment (garethellis0) [inline src/software/backend/grsim_backend.cpp]: A note to anyone reviewing: there's some trickyness with the factories here that we will need to figure out, closely related to https://github.com/UBC-Thunderbots/Software/pull/1163#issuecomment-586733276

File: src/software/visualizer/widgets/ai_control.cpp
Comment (garethellis0) [inline src/software/visualizer/widgets/ai_control.cpp]: We could probably clean this up a bit by adding `auto refbox_config = config->getMutableAIControlConfig()->getMutableRefboxConfig()->getMutableRefboxConfig();`  and then using that, but totally your call, it's not too bad as-is.

File: src/software/visualizer/widgets/ai_control.cpp
Comment (garethellis0) [inline src/software/visualizer/widgets/ai_control.cpp]: Similar comment here, could add `auto refbox_config = ...` at the top of this lambda and clean it up a bit, but your call.

File: src/software/visualizer/widgets/main_widget.cpp
Comment (garethellis0) [reply src/software/visualizer/widgets/main_widget.cpp:5]: Looks like it's passed into `setupAIControls` on line 36?

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Looking good! Really happy to see proper usage of the new config structure.

File: src/software/ai/hl/stp/stp.h
Comment (kelvinkoon) [inline src/software/ai/hl/stp/stp.h]: Also should be control_config here and subsequent areas.

File: src/software/backend/input/network/networking/network_filter.cpp
Comment (kelvinkoon) [inline src/software/backend/input/network/networking/network_filter.cpp]: Use refbox_config to keep it consistent.

**Review by kelvinkoon — COMMENTED** [review]
LGTM besides specifying the config names and pending changes.

File: src/software/visualizer/widgets/main_widget.cpp
Comment (HaomiaoZ) [reply src/software/visualizer/widgets/main_widget.cpp:5]: Yes you are right, didn't see that

File: src/software/backend/input/network/networking/network_filter.cpp
Comment (LiCody) [reply src/software/backend/input/network/networking/network_filter.cpp]: changed

File: src/software/ai/hl/stp/stp.h
Comment (LiCody) [reply src/software/ai/hl/stp/stp.h]: changed

File: src/software/ai/hl/stp/stp.cpp
Comment (LiCody) [reply src/software/ai/hl/stp/stp.cpp]: changed

File: src/software/visualizer/widgets/ai_control.cpp
Comment (LiCody) [reply src/software/visualizer/widgets/ai_control.cpp]: done

File: src/software/visualizer/widgets/ai_control.cpp
Comment (LiCody) [reply src/software/visualizer/widgets/ai_control.cpp]: done

File: src/software/ai/hl/stp/stp.cpp
Comment (LiCody) [reply src/software/ai/hl/stp/stp.cpp]: removed

File: src/software/ai/ai.h
Comment (LiCody) [reply src/software/ai/ai.h:25]: updated

**Review by akhilveeraghanta — COMMENTED** [review]
LGTM overall, but see if you can run `full_system_main` and make sure its not segfaulting

**Review by MathewMacDougall — DISMISSED** [review]
LGTM as long as it doesn't segfault

File: src/software/visualizer/widgets/ai_control.cpp
Comment (garethellis0) [reply src/software/visualizer/widgets/ai_control.cpp]: Just realized there's one more case of this above around line 40. Can we change that as well to be consistent?

**Comment by garethellis0** [comment]
Apologies for the delay in re-reviewing. Looking pretty good, just two comments left from me.

File: src/software/visualizer/widgets/ai_control.cpp
Comment (LiCody) [reply src/software/visualizer/widgets/ai_control.cpp]: changed

**Review by MathewMacDougall — DISMISSED** [review]

**Comment by LiCody** [comment]
It seems to be segfaulting at ```QApplication* application = new QApplication(argc, argv);``` in ```visualizer_wrapper.cpp``` when i try to run full system main with flags --backend grsim and im not sure why

**Comment by garethellis0** [comment]
Ah, this is a known issue that should be resolved by https://github.com/UBC-Thunderbots/Software/pull/1177. In the meantime, can you try passing the `--headless` flag to `full_system_main` when you run it? That should disable the GUI.

**Comment by LiCody** [comment]
it seems to run fine headless

**Review by MathewMacDougall — APPROVED** [review]

**Review by kelvinkoon — APPROVED** [review]

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Review by HaomiaoZ — APPROVED** [review]
LGTM 👍

**Pull request merged**
